ci: Prepare to add esp32 to size report.

jepler opened the pull request:
### Summary

I noticed that no xtensa builds were reported in the size report. I also made improvements to the surrounding scripts:
 * metrics.py now uses all available CPU parallelsim for calling "make"
 * metrics.py now cleans mpy-cross if it's used
 * for local testing, the set of PORTS_TO_CHECK, REFERENCE and COMPARISON can be overridden
 * for local testing, the original branch is restored after ci_code_size_build
 * pipefail manipulation is simplified by use of subshells
 * the command to generate the size change report is now a ci.sh step, ci_code_size_report
 * a possible unrelated error due to failure within the pipeline `zcat | tar` will now be caught
 * metrics.py can take a "pre_cmd" for a given port, used to inject the idf shell environment when building that port. A port's pre_cmd can be added or changed by setting an environment variable.

### Testing

I ran various ci.sh steps locally.

### Trade-offs and Alternatives

My past experience with having the idf environment loaded while trying to do other-port building made me believe that the approach I took is preferable to having the idf environment active through all of the size check build.

Due to the fact that the git version of `metrics.py` is used, _it breaks to actually add the build to the size check as part of this PR_, because at the base ref the build fails.

```diff
From fd5bf462c04c4135886d364a624333474fd9e170 Mon Sep 17 00:00:00 2001
From: Jeff Epler <jepler@gmail.com>
Date: Wed, 3 Sep 2025 13:49:55 -0500
Subject: [PATCH] ci: Add esp32 to code size build.

Signed-off-by: Jeff Epler <jepler@gmail.com>
---
 tools/ci.sh | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/tools/ci.sh b/tools/ci.sh
index 5417ce8797..bfd1aaf357 100755
--- a/tools/ci.sh
+++ b/tools/ci.sh
@@ -75,11 +75,12 @@ function ci_code_size_setup {
     ci_gcc_arm_setup
     ci_gcc_riscv_setup
     ci_picotool_setup
+    ci_esp32_idf_setup
 }
 
 function ci_code_size_build {
     # check the following ports for the change in their code size
-    PORTS_TO_CHECK=${1-bmusxpdv}
+    PORTS_TO_CHECK=${1-bmusxpdv3}
     SUBMODULES="lib/asf4 lib/berkeley-db-1.xx lib/btstack lib/cyw43-driver lib/lwip lib/mbedtls lib/micropython-lib lib/nxp_driver lib/pico-sdk lib/stm32lib lib/tinyusb"
 
     # Default GitHub pull request sets HEAD to a generated merge commit
-- 
2.47.2

```

jepler committed:
metrics: Use all CPUs during comparison builds.

Don't simply hard-code a parallelism of 3 for the build.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
ci: Let code_size_build be customized via environment.

For instance, to compare HEAD to origin/master on only the minimal
x86 build, use
```
PORTS_TO_CHECK=m REFERENCE=origin/master tools/ci.sh code_size_build
```

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
ci: Manipulate pipefail better.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
ci: Add, use ci_code_size_report.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
ci: Exit with error if zcat fails during ci_esp8266_setup.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
metrics: When cleaning, clean mpy-cross.

Changes of interest could be in emitted mpy frozen files.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
metrics: Allow pre_cmd to set up environment.

This is a necessary step to allow cleanly building the xtensa
port during CI, as it's undesirable to apply the esp-idf environment
settings while building other ports.

PRE_CMD_<letter> can be used to override the command, or add a command
if not otherwise specified. Note that <letter> is case sensitive.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

metrics: allow overriding pre_cmd

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
ci: Return to old branch after code_size_build.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

dpgeorge commented on tools/ci.sh:
```diff
@@ -93,33 +93,38 @@ function ci_code_size_build {
-    set +o pipefail
-    unset -f code_size_build_step
+        # build reference, save to size0
+        # ignore any errors with this build, in case master is failing
```
Is this comment still true, are errors in the REFERENCE build still ignored?

jepler replied:
I restored it, and tested that REFERENCE failures don't fail the overall script but COMPARISON failures do. This was "easy" since I could use a broken ref for REFERENCE or COMPARISON via environment settings.

dpgeorge replied:
OK, thanks.

dpgeorge commented on tools/ci.sh:
```diff
@@ -93,33 +93,38 @@ function ci_code_size_build {
+        # ignore any errors with this build, in case master is failing
+        code_size_build_step $REFERENCE ~/size0
+        # build PR/branch, save to size1
+        code_size_build_step $COMPARISON ~/size1
```
Is the return value of this last step propagated out as the return value / result of the entire `ci_code_size_build` function?

jepler replied:
We're running in `set -e` so any individual nonzero exit status not otherwise handled makes the whole thing exit. So if build_step fails (whether for REFERENCE or COMPARISON) the whole job fails.

In my local WIP I've restored $IGNORE_ERRORS, which means that a failure at REFERENCE results in success (zero exit status), while a failure at COMPARISON results in success (nonzero exit status)

```
(micropython) jepler@bert:~/src/micropython$ PORTS_TO_CHECK=bm REFERENCE=HEAD COMPARISON=broken-minimal bash -x ./tools/ci.sh code_size_build > /dev/null 2>&1 ; echo $?
1
(micropython) jepler@bert:~/src/micropython$ PORTS_TO_CHECK=bm REFERENCE=broken-minimal COMPARISON=HEAD bash -x ./tools/ci.sh code_size_build > /dev/null 2>&1 ; echo $?
0
```

dpgeorge commented on tools/ci.sh:
```diff
@@ -93,33 +93,38 @@ function ci_code_size_build {
 
-    return $STATUS
+function ci_code_size_report {
+    # Allow errors from tools/metrics.py to propagate out of the pipe above.
```
above -> below?

dpgeorge commented on tools/metrics.py:
```diff
@@ -65,7 +66,12 @@ def __init__(self, name, dir, output, make_flags=None):
+        "esp32",
+        "esp32",
+        "build-ESP32_GENERIC/micropython.elf",
+        pre_make_cmd=[".", "esp-idf/export.sh"],
```
It feels like this should be configurable by a command-line option, or environment variable.  (Eg for myself I have a shell alias `idf` which gets me the correct IDF environment.)

jepler replied:
In my WIP I've implemented it so that an environment variable such as `PRE_CMD_b=fnord` will supply a different PRE_CMD for the build tagged with 'b'.

jepler replied:
now pushed.

dpgeorge commented on tools/metrics.py:
```diff
@@ -75,14 +81,22 @@ def __init__(self, name, dir, output, make_flags=None):
+    return " ".join(shlex.quote(word) for word in args)
+
+
+def syscmd(*args, pre_make_cmd=None):
```
Maybe it can be just `pre_cmd`, because this `syscmd` function can be used for more than just running make (in principle).

dpgeorge commented:
The main reason to not include esp32 in the size report is that it takes a long time to build, and the code-size-report step is already one of the slowest CI jobs.

Nevertheless, code size reports have proved very useful (to me at least), so probably it's a good idea to add esp32 as well.

jepler committed:
ci: Re-add IGNORE_ERRORS.

It was removed (possibly inadvertently)
in 4c55b0879b38b373b44e84552d6754b7842b5b72

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler commented on .github/workflows/code_size.yml:
```diff
@@ -33,7 +33,7 @@ jobs:
       run: source tools/ci.sh && ci_code_size_build
     - name: Compute code size difference
-      run: tools/metrics.py diff ~/size0 ~/size1 | tee diff
+      run: source tools/ci.sh && ci_code_size_report
```
should the CI scripts be changed to use the `tools/ci.sh code_size_report` style instead (as a separate PR)?

jepler commented on tools/ci.sh:
```diff
@@ -82,47 +82,58 @@ function ci_code_size_setup {
+    # Override the list by setting PORTS_TO_CHECK in the environment before invoking ci.
+    : ${PORTS_TO_CHECK:=bmusxpdv}
+    
     SUBMODULES="lib/asf4 lib/berkeley-db-1.xx lib/btstack lib/cyw43-driver lib/lwip lib/mbedtls lib/micropython-lib lib/nxp_driver lib/pico-sdk lib/stm32lib lib/tinyusb"
```
should this be moved into the metrics script, like `metrics.py submodule  "$PORTS_TO_CHECK"`? (which would "make submodules" in each location called for)

projectgus reviewed (approved):
This change looks good to me, thanks @jepler!

Regarding performance, we have some caching steps in `ports_esp32.yml` ("Cached ESP-IDF install" and "ccache") that could be worth bringing into the code_size GitHub workflow as well to save build time (I'm not sure if there's a good way we can reuse them, or if copy/paste is the only way.)

(If we wanted to really speed code_size calculation up, then I think we could use the "GitHub Artifacts" feature to temporarily store the build directories from the other build jobs and then copy all their contents into the code_size job's working directory before it runs. If we then had a flag to let `metrics.py` do partial builds rather than clean ones then it'd be very fast to complete. But this is still probably a non-trivial amount of messing around.)

dpgeorge commented:
Rebased and merged in fef414eca49c44180fba9a0ab32e647c0373e67f through 13d49751f4f656c4ea12f3ed4cde2c404cf557ae

projectgus commented:
It's a typo. `s/pipefile/pipefail/`. One sec, will submit a PR.

projectgus commented:
I don't know what's going on with the second one, though...

projectgus commented:
The esp32 thing seems to be a broken package update from Espressif, not antyhing we've changed. Have linked the IDF issue from #18225.
